### PR TITLE
feat: add API for changing vfolder ownership

### LINKED ISSUE #408: vfolder migration between users and projects
In the current architecture, we could just change the owner field of the vfolders table as we use a flat UUID namespace for all vfolders.

Still, we need to take extra care when vfolders should be migrated to different storage proxy volumes as it requires explicit data copy and quota checks.

When we introduce a new planned vfolder architecture to better support per-user/per-group quota, we also need to let the storage-proxy to migrate vfolders between different quota namespaces and _depending on the storage vendors_ this may or may not require explicit data copy.

Let's abstract the vfolder migration operation in the storage-proxy as a long-running background process and make the manager API to proxy the progress with necessary database updates.

Q. Should we introduce a "locked" status of vfolders which prevent mounting of vfolders during such long-running operation?  When locking a vfolder, we also need to check existing mounts, but currently there is no database index for vfolder mounts.
Q. During the migration, the storage space will be taken twice of the vfolder size if we copy the full content first and then delete the original vfolder.  Should we incrementally migrate individual files to limit additional space usage?  If so, how to handle failure in the middle?

### PR DESCRIPTION
**Pull request by lizable**
This PR partially resolves #408.

FYI: 
- Support changing the `owner` of vfolder, not `creator`.
- Support only users who can access the same storage host that vfolder is located. (It will be expanded in phase 2)
- Support changing `user` type of vfolder, not `group(project)` type. (This also will be implemented in phase 2)

### CODE REVIEW

File: src/ai/backend/manager/api/vfolder.py [hunk 1]
```diff
@@ -3055,6 +3127,7 @@ def create_app(default_cors_options):
     cors.add(add_route("GET", r"/_/mounts", list_mounts))
     cors.add(add_route("POST", r"/_/mounts", mount_host))
     cors.add(add_route("DELETE", r"/_/mounts", umount_host))
+    cors.add(add_route("POST", r"/_/change_ownership", change_vfolder_ownership))
```
Comment (kyujin-cho) [inline src/ai/backend/manager/api/vfolder.py]: Using underbar to represent whitespace is now considered as legacy naming convention. Please use hyphen.
```suggestion
    cors.add(add_route("POST", r"/_/change-ownership", change_vfolder_ownership))
```

File: src/ai/backend/client/func/vfolder.py [hunk 2]
```diff
@@ -621,3 +621,16 @@ async def update_shared_vfolder(cls, vfolder: str, user: str, perm: str = None):
         )
         async with rqst.fetch() as resp:
             return await resp.json()
+
+    @api_function
+    @classmethod
+    async def change_vfolder_ownership(cls, vfolder: str, user: str):
+        rqst = Request("POST", "/folders/_/change_ownership")
```
Comment (kyujin-cho) [inline src/ai/backend/client/func/vfolder.py]: Also here. 
```suggestion
        rqst = Request("POST", "/folders/_/change-ownership")
```

File: src/ai/backend/manager/api/vfolder.py [hunk 3]
```diff
@@ -2985,6 +2986,77 @@ async def storage_task_exception_handler(
...
+        #       for now we only support migration btw user folder only
+        #
+        query = (
+            sa.update(vfolders)
+            .values(user=user_uuid)
+            .where(vfolders.c.id == vfolder_id)
+            .where(vfolders.c.ownership_type == VFolderOwnershipType.USER)
+        )
+        await conn.execute(query)
+    return web.json_response({}, status=200)
```
Comment (fregataa) [inline src/ai/backend/manager/api/vfolder.py:3073]:
```suggestion
    return web.Response(status=200)
```
we can just return status with `web.Response`

**Review by kyujin-cho — CHANGES_REQUESTED** [review]

File: src/ai/backend/manager/api/vfolder.py [hunk 4]
```diff
@@ -2985,6 +2986,77 @@ async def storage_task_exception_handler(
...
+        user_uuid,
+    )
+    allowed_hosts_by_user = VFolderHostPermissionMap()
+    async with root_ctx.db.begin_readonly() as conn:
+        j = sa.join(users, keypairs, users.c.uuid == keypairs.c.user)
+        query = (
+            sa.select([users.c.email, users.c.domain_name, keypairs.c.resource_policy])
+            .select_from(j)
+            .where(users.c.uuid == user_uuid)
+            .where(users.c.status == users.UserStatus.ACTIVE)
```
Comment (fregataa) [inline src/ai/backend/manager/api/vfolder.py]:
```suggestion
            .where(users.c.uuid == user_uuid & users.c.status == users.UserStatus.ACTIVE)
```

File: src/ai/backend/manager/api/vfolder.py [hunk 5]
```diff
@@ -2985,6 +2986,77 @@ async def storage_task_exception_handler(
...
+    For now, we only provide changing the ownership of user-folder
+    """
+    access_key = request["keypair"]["access_key"]
+    vfolder_id = params["vfolder"]
+    user_uuid = params["user"]
+    root_ctx: RootContext = request.app["_root.context"]
+
+    # TODO: Implement Change ownership using DB transaction
+    log.info(
+        "VFOLDER.CHANGE_VFOLDER_OWNERSHIP(emial:{}, ak:{}, vfid:{}, uid:{})",
```
Comment (fregataa) [inline src/ai/backend/manager/api/vfolder.py]:
```suggestion
        "VFOLDER.CHANGE_VFOLDER_OWNERSHIP(email:{}, ak:{}, vfid:{}, uid:{})",
```
mistypo on `email`

File: src/ai/backend/manager/api/vfolder.py [hunk 6]
```diff
@@ -2985,6 +2986,77 @@ async def storage_task_exception_handler(
...
+        # TODO: we need to implement migration from project to other project
+        #       for now we only support migration btw user folder only
+        #
+        query = (
+            sa.update(vfolders)
+            .values(user=user_uuid)
+            .where(vfolders.c.id == vfolder_id)
+            .where(vfolders.c.ownership_type == VFolderOwnershipType.USER)
+        )
+        await conn.execute(query)
```
Comment (fregataa) [inline src/ai/backend/manager/api/vfolder.py]:
```suggestion
    async def _update() -> None:
        async with root_ctx.db.begin() as conn:
        # TODO: we need to implement migration from project to other project
        #       for now we only support migration btw user folder only
        #
            query = (
                sa.update(vfolders)
                .values(user=user_uuid)
                .where(vfolders.c.id == vfolder_id & vfolders.c.ownership_type == VFolderOwnershipType.USER)
            )
            await conn.execute(query)
    await execute_with_retry(_update)
```
Any insert or update queries should be executed in `execute_with_retry` function.

**Review by kyujin-cho — APPROVED** [review]

**Review by fregataa — CHANGES_REQUESTED** [review]
Thank you for your work on it. I left some comments

File: src/ai/backend/manager/api/vfolder.py (hunk 5 above)
Comment (lizable) [reply src/ai/backend/manager/api/vfolder.py]: Bullseye! 👀

File: src/ai/backend/manager/api/vfolder.py (hunk 6 above)
Comment (lizable) [reply src/ai/backend/manager/api/vfolder.py]: Noted!

File: src/ai/backend/manager/api/vfolder.py (hunk 3 above)
Comment (lizable) [reply src/ai/backend/manager/api/vfolder.py:3073]: I agree. I'll apply it right away.

File: src/ai/backend/manager/api/vfolder.py (hunk 4 above)
Comment (lizable) [reply src/ai/backend/manager/api/vfolder.py]: Noted!

**Review by kyujin-cho — APPROVED** [review]

**Pull request merged**